client/descriptions.go

```diff
@@ -312,19 +331,54 @@ func (f FieldID) String() string {
 	return fmt.Sprint(uint32(f))
 }
 
-// FieldDescription describes a field on a Schema and its associated metadata.
-type FieldDescription struct {
+// SchemaFieldDescription describes a field on a Schema and its associated metadata.
+type SchemaFieldDescription struct {
 	// Name contains the name of this field.
 	//
 	// It is currently immutable.
 	Name string
 
-	// ID contains the internal ID of this field.
+	// The data type that this field holds.
 	//
-	// Whilst this ID will typically match the field's index within the Schema's Fields
-	// slice, there is no guarantee that they will be the same.
+	// Must contain a valid value. It is currently immutable.
+	Kind FieldKind
+
+	// Schema contains the schema name of the type this field contains if this field is
+	// a relation field.  Otherwise this will be empty.
+	Schema string
+
+	// RelationName the name of the relationship that this field represents if this field is
+	// a relation field.  Otherwise this will be empty.
+	RelationName string
+
+	// The CRDT Type of this field. If no type has been provided it will default to [LWW_REGISTER].
 	//
-	// It is immutable.
+	// It is currently immutable.
+	Typ CType
+
+	// If true, this is the primary half of a relation, otherwise is false.
+	IsPrimaryRelation bool
+}
+
+// CollectionFieldDescription describes the local components of a field on a collection.
+type CollectionFieldDescription struct {
+	// Name contains the name of the [SchemaFieldDescription] that this field uses.
+	Name string
+
+	// ID contains the local, internal ID of this field.
+	ID FieldID
+}
+
+// FieldDescription describes the combined local and global set of properties that constitutes
+// a field on a collection.
+//
+// It draws it's information from the [CollectionFieldDescription] on the [CollectionDescription],
+// and the [SchemaFieldDescription] on the [SchemaDescription].
+type FieldDescription struct {
```

Comment:
suggestion: It really threw me off to see `FieldDescription` still in existance. It's an exact duplicate of `SchemaFieldDescription` with tha addition of the `ID` field. Every function that returns `FieldDescription` or uses it is a method of `CollectionDescription` or `collection` and it's odd that when `FieldDescription` is returned you have to construct the struct with `NewFieldDescription`. A much cleaner approach would be to embed the `SchemaFieldDescription` into the `CollectionFieldDescription`.

Reply:
> It really threw me off to see FieldDescription still in existance.

That's not good - do you have a rough idea as to how much of that was caused my terrible setting of expectations in discord, vs the actual change?

> Every function that returns FieldDescription or uses it is a method of CollectionDescription

Yes, that kind of felt like it was breaking expectations when I was consuming it.  It kind of flows better at the moment IMO, but that is very possibly because of the limited number of properties on collection-field.

> A much cleaner approach would be to embed the SchemaFieldDescription into the CollectionFieldDescription

I do very strongly disagree with this.  It creates a large amount of ambiguity in the data model and is kind of reverting back to the starting point where collection had a pointer to schema on it. Amongst other things, it would be very very confusing when mutating collection field stuff as to what you could and couldn't mutate - it is bad with a pointer, and much worse when embedding.

`CollectionFieldDescription` does not have a `SchemaFieldDescription`, and in the possibly quite near future, not all `CollectionFieldDescription` will even have a corresponding `SchemaFieldDescription` (think secondary fields in a relation).

----

I did originally have a todo here about moving this struct to the `core` package, but decided it was nice having the funcs that return it as public.  Maybe this choice should be revisited, along with the fact that the functions returning it are hosted on `CollectionFieldDescription` (which feels uglier the longer I look at it, even though it is kind of convenient atm).

- [ ] Discuss further, maybe move stuff

Reply:
> That's not good - do you have a rough idea as to how much of that was caused my terrible setting of expectations in discord, vs the actual change?

 I though `FieldDescription` would become `SchemaFieldDescription` and `CollectionFieldDescription`. Not have the two new struct plus `FieldDescription`.
 
 > I do very strongly disagree with this. It creates a large amount of ambiguity in the data model
 
 It feels very ambiguous the way it is now in my opinion.  
 
 > is kind of reverting back to the starting point where collection had a pointer to schema on it
 
I disagree. Everywhere `FieldDescription` is used, both collection and schema description are needed to build the struct. It would be simple to remove the confusion and the have the same functionality you have now with something like this
 
```go
type CollectionFieldDescription struct {
  Name string
  ID FieldID

  schemaField immutable.Option[SchemaFieldDescription] `json:"-"`
```
}

Reply:
> I disagree. Everywhere FieldDescription is used, both collection and schema description are needed to build the struct. It would be simple to remove the confusion and the have the same functionality you have now with something like this

Only at the moment, because we have only moved one property.

Reply:
As far as I can tell it would only get worse with more fields in the `CollectionFieldDescription`.

Reply:
The suggestion is totally corrupting the purpose of `CollectionFieldDescription`.  It is no longer modelling just collection fields, and is instead becoming the kind of 'God class' that schema field is/was (and the old collection description with pointer).

There is a loss of separation in the model.  It is dirty and ambiguous.

Just because you dislike FieldDescription (essentially a utility type that could be made internal, it is not actually part of the model) in it's current form, does not mean the only alternative is to blend everything back together again on the public model.

Reply:
> The suggestion is totally corrupting the purpose of CollectionFieldDescription. It is no longer modelling just collection fields, and is instead becoming the kind of 'God class' that schema field is/was (and the old collection description with pointer).

I don't understand why you are saying this. Everywhere `FieldDescription` is needed, we need both `CollectionFieldDescription` and `SchemaDescription`. Also, everywhere it is used is in a collection context which means a `CollectionFieldDescription` makes a lot more sense here. Since you are saying that at some point a `CollectionFieldDescription` could exists without a relation to a `SchemaFieldDescription`, then it renders `FieldDescription` even more confusing in its usage. However, having an optional `SchemaFieldDescription` in the `CollectionFieldDescription` makes it a lot more obvious what is happening and removes the ambiguity that `FieldDescription` brings in collection contexts.

> There is a loss of separation in the model. It is dirty and ambiguous.

The separation is still there. The is only an optional value added when it makes sense to have it. Like I said above, much less ambiduous. Nothing dirty here.

> Just because you dislike FieldDescription

It's not that I dislike it, it's that it is extremely confusing to be using the `FieldDescription` in a collection context when a `CollectionFieldDescription` exists.

> The suggestion largely defeats the point of this PR and line of work.

It does not. You can still store the two separately and use `SchemaFieldDescription` and `CollectionFieldDescription` on their own. But when `CollectionFieldDescription` has a related `SchemaFieldDescription` that need to be present for different manipulations, we won't need to copy all the field to a new type just for that exact purpose.

Reply:
If you did something like this https://go.dev/play/p/SwCupd3Ep2C it would eliminate the duplicated fields and methods. Doesn't remove the ambiguity in the collection context though.

Reply:
I had that originally, I got rid of it because `Name` is duplicated/ambiguous, and that problem is moderately likely to happen with a few other props in the future.

Having to paste a couple of new props every now and then from Collection/SchemaFieldDescription onto FieldDescription is not worth dealing with that mess.

Reply:
I have pushed changes that I am happy with.  I realized part way through that (what was) `FieldDescription` is basically what `CollectionDefinition` is, but at a field level.  So I ended up just cleaning up the `GetFieldByName` funcs and renaming it.

`CollectionDefinition` will eventually suffer from the same (and worse) ambiguity that `FieldDescription.Name` does when embedding, however at the moment we were able to get away with it because there is currently no overlap in properties.  This is almost certainly going to change, and `CollectionDefinition` will need to be changed for that (redeclaration of it's components in a way that eliminates ambiguity over what should be used, for example a single `Fields []FieldDefinition` property, containing a merged set of collection and schema fields).

Reply:
> and that problem is moderately likely to happen with a few other props in the future

To expand on this, I see the following properties likely to clash, and they may have different values on schema vs collection:

- `Name` (if/when it becomes a label this will be much worse, where it's value may be different on schema vs collection)
- `Description`, I definitely want this on Collection, and there is value on having one on Schema too (overridable by collection
- `RelationName`, I want this on Collection, but we could also allow it to be defined on Schema in the case where users wish to include both halves of a relation on the global schema.  In such a case I'm not sure we'd want to allow different values to be set, but we could also allow it to be overrideable like `Description` and `Name`.
- Potentially ACP field level stuff too, I can imaging people might want to be able to set field level stuff on both the schema, and allow local collections to define additional rules on top of that.

Reply:
Islam (https://github.com/sourcenetwork/defradb/pull/2339#discussion_r1503957788):

todo: I feel like it's wrong having almost identical structs with `SchemaFieldDescription`.

Moreover, you draw an analogy with collectin/schema description, but these 2 are embedded into `CollectionDefinition`.
But this one is not. It's inconsistent. I you think that `CollectionFieldDescription` is too small and has only 1 unique field `ID` we can make it:
```go
type FieldDefinition struct {
    schemaField SchemaFieldDescription
    collectionFieldID int
}
```
And you won't need to copy&past all method below (`IsObject`, `IsObjectArray`, ...)

Reply:
> Moreover, you draw an analogy with collectin/schema description, but these 2 are embedded into CollectionDefinition.

They shouldn't be long term, please see my earlier comment:

> This is almost certainly going to change, and CollectionDefinition will need to be changed for that (redeclaration of it's components in a way that eliminates ambiguity over what should be used, for example a single Fields []FieldDefinition property, containing a merged set of collection and schema fields).

> And you won't need to copy&past all method below (IsObject, IsObjectArray, ...)

I consider copy-paste here to be significantly cheaper than detangling inheritance hierarchies, especially coupled with how embedding structs in Go works (is half flattened, but not quite, so the end interface is not the same as declaring the properties in place).  And as mentioned further up, these property sets atm are very similar because only one has been moved, they will diverge more so in the future.

Reply:
> And you won't need to copy&past all method below (IsObject, IsObjectArray, ...)

Discussion point bypassed by Islam's suggestion of moving these functions onto FieldKind.